provider/plugin.go

```diff
@@ -33,55 +35,43 @@ func Serve() {
 	})
 }
 
-// ServeTest is for serving the provider in-process when testing
-// returns the reattach configuration or an error
-func ServeTest() (tfexec.ReattachInfo, error) {
+// ServeReattach is the entrypoint for manually starting the provider
+// as a process in reattach mode for debugging.
+func ServeReattach() {
 	reattachConfigCh := make(chan *plugin.ReattachConfig)
-
 	go func() {
-		plugin.Serve(&plugin.ServeConfig{
-			HandshakeConfig: handshakeConfig,
-			GRPCServer:      plugin.DefaultGRPCServer,
-			Plugins: plugin.PluginSet{
-				"provider": &grpcPlugin{
-					providerServer: &RawProviderServer{},
-				},
-			},
-			Test: &plugin.ServeTestConfig{
-				ReattachConfigCh: reattachConfigCh,
-			},
-			Logger: hclog.FromStandardLogger(Dlog, &hclog.LoggerOptions{
-				JSONFormat: false,
-				Level:      hclog.Info,
-			}),
-		})
+		reattachConfig, err := waitForReattachConfig(reattachConfigCh)
+		if err != nil {
+			fmt.Printf("Error getting reattach config: %s\n", err)
+			return
+		}
+		printReattachConfig(reattachConfig)
 	}()
 
-	select {
-	case reattachConfig := <-reattachConfigCh:
-		return map[string]tfexec.ReattachConfig{
-			"hashicorp/kubernetes-alpha": {
-				Protocol: string(reattachConfig.Protocol),
-				Pid:      reattachConfig.Pid,
-				Test:     true,
-				Addr: tfexec.ReattachConfigAddr{
-					Network: reattachConfig.Addr.Network(),
-					String:  reattachConfig.Addr.String(),
-				},
+	plugin.Serve(&plugin.ServeConfig{
+		HandshakeConfig: handshakeConfig,
+		GRPCServer:      plugin.DefaultGRPCServer,
+		Plugins: plugin.PluginSet{
+			"provider": &grpcPlugin{
+				providerServer: &RawProviderServer{},
 			},
-		}, nil
-	case <-time.After(10 * time.Second):
-		return nil, fmt.Errorf("timeout when starting the provider")
-	}
+		},
+		Test: &plugin.ServeTestConfig{
+			ReattachConfigCh: reattachConfigCh,
+		},
+		Logger: hclog.FromStandardLogger(Dlog, &hclog.LoggerOptions{
+			JSONFormat: false,
+			Level:      hclog.Debug,
+		}),
+	})
 }
 
-// DebugServe is the provider entry point when running in stand-alone process mode
-func DebugServe() {
-	reattachCh := make(chan *plugin.ReattachConfig)
-
-	go waitForReattachConfig(reattachCh)
+// ServeTest is for serving the provider in-process when testing.
+// Returns a ReattachInfo or an error.
+func ServeTest(ctx context.Context) (tfexec.ReattachInfo, error) {
+	reattachConfigCh := make(chan *plugin.ReattachConfig)
 
-	plugin.Serve(&plugin.ServeConfig{
+	go plugin.Serve(&plugin.ServeConfig{
```

Comment:
Doing this as a go routine without a controlling channel or some some other mechanism to keep track of it is a bit problematic. 
Some call this a go routine leak.

Without a controlling mechanism, as ServeAttach returns and in turn main returns after that, the main go routine is now finished and the process is in limbo because the other go routine that is launched here might still be running.

I don't really see a need to launch this as go routine, but if you want to you should also set ServeTestConfig.CloseCh to a channel passed down from main() where main waits for that channel to close, signaling that the server go routine has completed.

Reply:
This function will only bee called in `TestMain()`, where the plugin will have to be started in a separate go routine. We pass a Context into the `Test` parameter here, which can be used to stop it.

Reply:
But ServeReattach does the same thing with a go routine and that's called from the provider's own main().

Reply:
Not sure I'm understanding you. `ServeReattach()` uses the goroutine to wait for the reattach config then print it to stdout - which is fine to happen asynchronously because we're just spitting it out for the user to copy and paste. In `TestMain()` I need to start the provider and then block until the reattach info is received so that I can pass it onto the test helper after the provider is started.

Reply:
ServeReattach launches the whole plugin.Serve(...) in a go routine via an anonymous func. That's the bit I'm concerned about.
If you just want to wait for the reattach config you can call the waiter in a goroutine instead as that has a more determined lifecycle and completes once config is printed or time runs out. That way the main go routine is the longest running one.

Reply:
`ServeReattach` doesn't launch the whole thing in a goroutine - just the waiter part that's inside the goroutine anonymous func.

Reply:
Yeah, you're right. This is proof enough I shouldn't be doing code reviews on an iPhone anymore 👎 
I re-checked now on a bigger screen. Looks good.